cputest/unaligned_stores.cpp

```diff
@@ -0,0 +1,204 @@
+            "ACTUAL:              EXPECTED:\n"
+            "%08X %08X    %08X %08X\n"
+            "%08X %08X    %08X %08X\n"
+            "%08X %08X    %08X %08X\n"
+            "%08X %08X    %08X %08X\n",
+            size, reinterpret_cast<uintptr_t>(ptr + i),
+            Read(ptr), Read(ptr + 4), Read(reference_buffer), Read(reference_buffer + 4),
+            Read(ptr + 8), Read(ptr + 12), Read(reference_buffer + 8), Read(reference_buffer + 12),
+            Read(ptr + 16), Read(ptr + 20), Read(reference_buffer + 16), Read(reference_buffer + 20),
+            Read(ptr + 24), Read(ptr + 28), Read(reference_buffer + 24), Read(reference_buffer + 28));
+
+    // The pattern behind when PI ERROR interrupts are generated is not understood.
+    // When pi_error_expected == false, a real Wii does not generate an interrupt, but when
+    // pi_error_expected == true it may or may not. Disable this part of the test for now.
+    /*
+    int pi_error_expected = (i & 3) || size < 4;
+    int pi_error_actual = s_pi_error_occurred;
+
+    DO_TEST(pi_error_actual == pi_error_expected,
+            "%u-byte write to 0x%" PRIxPTR " failed\n"
+            "ACTUAL:              EXPECTED:\n"
+            "pi_error_occurred=%d  pi_error_occurred=%d\n",
+            size, reinterpret_cast<uintptr_t>(ptr + i),
+            pi_error_actual, pi_error_expected);
+    */
+  }
+}
+
+static void PIErrorHandler(u32 nIrq, void* pCtx)
+{
+  // Clear the interrupt
+  s_pi_reg[0] = 0x01;
+
+  s_pi_error_occurred = true;
+}
+
+static void Initialize()
+{
+  network_init();
+
+	// Initialise the video system
+	VIDEO_Init();
+
+	// This function initialises the attached controllers
+	WPAD_Init();
+
+	// Obtain the preferred video mode from the system
+	// This will correspond to the settings in the Wii menu
+	GXRModeObj* rmode = VIDEO_GetPreferredMode(NULL);
+
+	// Allocate memory for the display in the uncached region
+	void* xfb = MEM_K0_TO_K1(SYS_AllocateFramebuffer(rmode));
+
+	// Initialise the console, required for printf
+	console_init(xfb, 20, 20, rmode->fbWidth, rmode->xfbHeight, rmode->fbWidth * VI_DISPLAY_PIX_SZ);
+
+	// Set up the video registers with the chosen mode
+	VIDEO_Configure(rmode);
+
+	// Tell the video hardware where our display memory is
+	VIDEO_SetNextFramebuffer(xfb);
+
+	// Make the display visible
+	VIDEO_SetBlack(FALSE);
+
+	// Flush the video register changes to the hardware
+	VIDEO_Flush();
+
+	// Wait for Video setup to complete
+	VIDEO_WaitVSync();
+	if (rmode->viTVMode & VI_NON_INTERLACE)
+    VIDEO_WaitVSync();
+}
+
+int main()
+{
+  Initialize();
+
+  network_printf("Setting up PI ERROR interrupt handler...\n");
+  IRQ_Request(IRQ_PI_ERROR, PIErrorHandler, nullptr);
+  __UnmaskIrq(IM_PI_ERROR);
+
+  // Get a pointer to a page boundary, with at least 32 bytes available on each side of the boundary
+  network_printf("Allocating memory...\n");
+  u8* memory_allocation = new u8[PAGE_SIZE + 32 * 4];
```

Comment:
This doesn't really guarantee whether your buffer ends up in MEM1 or MEM2, which seems relevant for this test. I think libogc has allocation functions which actually do provide these guarantees?

Reply:
I've added a second commit that does this.